cctp_primitives/src/proving_system/verifier/batch_verifier.rs

```diff
@@ -0,0 +1,447 @@
+            let res = Self::batch_verify_proofs(
+                to_verify, g1_ck.as_ref().unwrap(),
+                g2_ck.as_ref().unwrap(), rng
+            );
+
+            // Return the id of the first failing proof if it's possible to determine it
+            if res.is_err() {
+                match res.unwrap_err() {
+                    Some(idx) => return Err(ProvingSystemError::FailedBatchVerification(Some(ids[idx].clone()))),
+                    None => return Err(ProvingSystemError::FailedBatchVerification(None))
+                }
+            }
+
+            Ok(res.unwrap())
+        }
+    }
+
+    /// Verify all the proofs in `verifier_data`.
+    /// If the verification procedure fails, it may be possible to get the id of
+    /// the proof that has caused the failure.
+    pub fn batch_verify_all<R: RngCore>(
+        &self,
+        rng: &mut R
+    ) -> Result<bool, ProvingSystemError>
+    {
+        self.batch_verify_subset(self.verifier_data.keys().map(|k| k.clone()).collect::<Vec<_>>(), rng)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use algebra::UniformRand;
+    use proof_systems::darlin::tests::{
+        simple_marlin::generate_test_data as generate_simple_marlin_test_data,
+        final_darlin::generate_test_data as generate_final_darlin_test_data,
+    };
+    use crate::{
+        proving_system::{
+            init::{load_g1_committer_key, get_g1_committer_key, load_g2_committer_key, get_g2_committer_key},
+            error::ProvingSystemError,
+            verifier::{UserInputs, ZendooVerifier, RawVerifierData},
+        },
+        type_mapping::{FieldElement, G1, G2},
+        utils::serialization::SerializationUtils
+    };
+    use poly_commit::ipa_pc::UniversalParams;
+    use rand::{thread_rng, Rng};
+    use serial_test::serial;
+
+    struct TestCircuitInputs {
+        c: FieldElement,
+        d: FieldElement
+    }
+
+    impl UserInputs for TestCircuitInputs {
+        fn get_circuit_inputs(&self) -> Result<Vec<FieldElement>, ProvingSystemError>
+        {
+            Ok(vec![self.c, self.d])
+        }
+    }
+
+    struct TestCircuitVerifier {}
+
+    impl ZendooVerifier for TestCircuitVerifier { type Inputs = TestCircuitInputs; }
+
+    fn get_params() -> (
+        UniversalParams<G1>,
+        UniversalParams<G2>,
+        usize,
+        usize,
+        String,
+        String,
+    ) {
+
+        let max_pow = 7usize;
+        let segment_size = 1 << max_pow;
+
+        // Init committer keys
+        let g1_ck_path = "./ck_g1";
+        let committer_key_g1 = {
+            load_g1_committer_key(segment_size - 1, g1_ck_path).unwrap();
+            get_g1_committer_key().unwrap()
+        }.as_ref().unwrap().clone();
+
+        let params_g1 = UniversalParams::<G1> {
+            comm_key: committer_key_g1.comm_key.clone(),
+            h: committer_key_g1.h.clone(),
+            s: committer_key_g1.s.clone(),
+        };
+
+        let g2_ck_path = "./ck_g2";
```

Comment:
Consider using a temporary path (such as the portable std::env::temp_dir()) for writing/deleting test files